<?xml version="1.0"?>
<launch>
  <arg name="use_sim_time" default="true" />
  <arg name="gui" default="true" />
  <arg name="headless" default="false" />

  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="debug" value="0" />
    <arg name="gui" value="$(arg gui)" />
    <arg name="use_sim_time" value="$(arg use_sim_time)" />
    <arg name="headless" value="$(arg headless)" />
    <arg name="paused" value="false"/>
  </include>

  <!-- Ideal formation position -->
  <rosparam param="formation_positions">[[0, 2, 0.5], [0, 0, 0.5], [0, -2, 0.5], [2, 2, 0.5], [2, -2, 0.5], [2, 0, 0.5], [-2, 0, 0.5]]</rosparam>

  <!-- CBF Communication maintenance or obstacle avoidance activation (1 is activated/0 is deactivated) -->
  <param name="cbf_cm" value="1"/>
  <param name="cbf_oa" value="1"/>

  <!-- Safe distance for communication maintenance and obstacle avoidance -->
  <param name="safe_distance_cm" value="1.3"/>
  <param name="safe_distance_oa" value="0.9"/>

  <!-- Nexus 1  -->
  <group ns="nexus1">
    <!-- Load robot description -->
    <include file="$(find sml_nexus_description)/launch/sml_nexus_description.launch" />
    <!-- Spawn the robot -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model"
      args="-urdf -model nexus1 -param robot_description -x -1 -y 3 -z 0.5" />
  </group>
    
  <!-- Nexus 2  -->
  <group ns="nexus2">
    <!-- Load robot description -->
    <include file="$(find sml_nexus_description)/launch/sml_nexus_description.launch" />
    <!-- Spawn the robot -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model"
      args="-urdf -model nexus2 -param robot_description -x 0 -y 0 -z 0.5" />
  </group>

  <!-- Nexus 3  -->
  <group ns="nexus3">
    <!-- Load robot description -->
    <include file="$(find sml_nexus_description)/launch/sml_nexus_description.launch" />
    <!-- Spawn the robot -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model"
      args="-urdf -model nexus3 -param robot_description -x 0 -y -3 -z 0.5" />
  </group>

  <!-- Nexus 4  -->
  <group ns="nexus4">
    <!-- Load robot description -->
    <include file="$(find sml_nexus_description)/launch/sml_nexus_description.launch" />
    <!-- Spawn the robot -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model"
      args="-urdf -model nexus4 -param robot_description -x 3 -y 3 -z 0.5" />
  </group>

  <!-- Nexus 5  -->
  <group ns="nexus5">
    <!-- Load robot description -->
    <include file="$(find sml_nexus_description)/launch/sml_nexus_description.launch" />
    <!-- Spawn the robot -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model"
      args="-urdf -model nexus5 -param robot_description -x 3 -y -2 -z 0.5" />
  </group>

  <!-- Nexus 6  -->
  <group ns="nexus6">
    <!-- Load robot description -->
    <include file="$(find sml_nexus_description)/launch/sml_nexus_description.launch" />
    <!-- Spawn the robot -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model"
      args="-urdf -model nexus6 -param robot_description -x 5 -y -1 -z 0.5" />
  </group>

  <!-- Nexus 7  -->
  <group ns="nexus7">
    <!-- Load robot description -->
    <include file="$(find sml_nexus_description)/launch/sml_nexus_description.launch" />
    <!-- Spawn the robot -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model"
      args="-urdf -model nexus7 -param robot_description -x -3 -y 1 -z 0.5" />
  </group>

  <!-- Centralized auto node -->
  <group ns="Auto">
    <node name="cbf_formation_controller_centralized" pkg="first_task_formation_cbf" type="formation_keeping_centralized.py" output="screen" >
      <rosparam param="robots_number">[1, 2, 3, 4, 5, 6, 7]</rosparam>
      <rosparam param="neighbours">[[2, 4], [1, 3, 4, 5, 6, 7], [2, 5], [1, 2, 6], [2, 3, 6], [2, 4, 5], [2]]</rosparam>
    </node>
  </group>

  <!-- Motion capture system simulation -->
  <include file="$(find mocap_simulator)/launch/qualisys_simulator.launch" />

</launch>
